### PR TITLE
webserver: expose and emit CockpitWebRequest

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
The handle-stream signal on CockpitWebServer has a veritable grab-bag of parameters.  I've wanted to extend it several times in the past, but have always shied away.
    
Let's start to clean this up.
    
Expose our internal CockpitRequest as a public type named CockpitWebRequest.  Instead of emitting six separate parameters on the handle-stream function, we now emit this single object.

We also push this object down through our stack of functions designed for responding to web requests, allowing for a large number of cleanups to move various bits of webserver implementation detail from the consumers of this API to be accessible via new accessors on CockpitWebRequest.

### CODE REVIEW

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! This looks by and large sensible to me, thanks for the primer during our meeting! I have a few little complaints, and this requires the obvious de-conflicting/rebasing/squashing/fixing commit messages.

File: src/common/cockpitwebserver.c [hunk 1]
```diff
@@ -69,11 +86,11 @@ enum
 static gint sig_handle_stream = 0;
 static gint sig_handle_resource = 0;
 
-static void cockpit_request_free (gpointer data);
+static void cockpit_web_request_free (gpointer data);
 
-static void cockpit_request_start (CockpitWebServer *self,
-                                   GIOStream *stream,
-                                   gboolean first);
+static void cockpit_web_request_start (CockpitWebServer *self,
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c:95]: I see it kinda as a static method on `CockpitWebRequest` which constructs (but doesn't return) an object of that type, and takes `CockpitWebServer` as a first parameter.

File: src/ws/cockpitauth.c [hunk 2]
```diff
@@ -374,11 +373,12 @@ cockpit_auth_steal_authorization (GHashTable *headers,
...
   gpointer key;
 
-  g_assert (headers != NULL);
+  g_assert (request != NULL);
   g_assert (ret_conversation != NULL);
   g_assert (ret_type != NULL);
 
   /* Avoid copying as it can contain passwords */
+  GHashTable *headers = cockpit_web_request_get_headers (request);
   if (g_hash_table_lookup_extended (headers, "Authorization", &key, (gpointer *)&line))
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/ws/cockpitauth.c:334]: OK.

File: src/ws/cockpitauth.c [hunk 3]
```diff
@@ -374,11 +373,12 @@ cockpit_auth_steal_authorization (GHashTable *headers,
   gchar *line;
   gpointer key;
 
-  g_assert (headers != NULL);
+  g_assert (request != NULL);
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/ws/cockpitauth.c:328]: Sounds good.

File: src/common/cockpitwebserver.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/common/cockpitwebserver.c:95]: Then let's at least rename `self` to `webserver`?

File: src/common/cockpitwebserver.c [hunk 4]
```diff
@@ -837,8 +822,13 @@ cockpit_web_request_process (CockpitWebRequest *self,
       return;
     }
 
-  self->original_path = path;
-  self->path = path + self->web_server->url_root->len;
+  self->original_path = g_strdup (path);
+  self->path = self->original_path + self->web_server->url_root->len;
+  self->query = strchr (self->path, '?');
+  if (self->query)
+    *((gchar *) self->query++) = '\0';
```
Comment (martinpitt) [inline src/common/cockpitwebserver.c:838]: As this gets modified, why not drop the `const` from the declaration in cockpitwebrequest-private.h and get rid of the cast here? But this is still a bit confusing, as this is now really just a pointer into the `path`, which is also `const gchar*`. But then, `path` by itself just points into `original_path`, which at last is writable. 

This is all a bit hard to follow, TBH. Would it be possible to flip this around, so that `original_path` is const, and actually never modified? and `path` gets strdup()'ed, truncated etc.? `query` should stay const I think, due to the `else` branch.

**Review by martinpitt — COMMENTED** [review]
Two borderline nitpicks, and one question. This is not an outright rejection, but let's discuss these quickly.

Current test failures are all unrelated, FTR.

File: src/common/cockpitwebserver.c [hunk 5]
```diff
@@ -319,23 +318,22 @@ cockpit_web_server_default_handle_stream (CockpitWebServer *self,
...
-      pos = strchr (request->path + 1, '/');
-      if (pos != NULL)
-        {
-          pos++;
-          bak = *pos;
-          *pos = '\0';
-        }
+      gchar buffer[component_end + 1];
+      memcpy (buffer, request->path, component_end);
+      buffer[component_end] = '\0';
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c:327]: You have no idea how many times I rewrote this, and how many different string functions I considered, including `strncpy()`, `stpncpy()` and even `g_strlcpy()`...

Neither of `strncpy()` nor `strpcpy()` will nul-terminate the output unless there is a nul terminator among the first characters of the input (which is usually not the case here).

Now that I think about it, though, `strlcpy()` probably actually works here.  I'll try changing it.

File: src/common/cockpitwebserver.c (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c:838]: This was another case of "...I rewrote this so many freaking times..."

That's a good suggestion, though.  I'll look into it.

File: src/common/cockpitwebserver.c [hunk 6]
```diff
@@ -319,23 +318,22 @@ cockpit_web_server_default_handle_stream (CockpitWebServer *self,
...
    */
-
-  /* Temporarily null terminate string after first component */
-  pos = NULL;
-  if (request->path[0] != '\0')
+  g_assert (request->path[0] == '/');
+  gsize component_end = 1 + strcspn (request->path + 1, "/");
+  if (request->path[component_end] == '/')
+    component_end++;
+  if (component_end < 100)
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c:323]: That's not what's happening here.

component_end is under control of the user.  The user might well specify a URL like `/reallyreallyreally[x200]long/` and we have no opinion about if that's valid or not.

This is about finding the handlers that were connected with detailed strings like `"handle-resource::/favicon.ico"`.  We need to find those handlers, if they exist, but we know that none of those will exist for more than 100 characters (or probably 20 or so, to be honest).  It's about saying "let's not bother looking for it if it's so big".

File: src/common/cockpitwebserver.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/common/cockpitwebserver.c:327]: Argh right, sorry about that. strlcpy() does sound nice indeed, it's what asprintf() does, except much cheaper.

File: src/common/cockpitwebserver.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/common/cockpitwebserver.c:323]: I see, thanks!

File: src/common/cockpitwebserver.c (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/common/cockpitwebserver.c:838]: I remember why I did this now: `path` gets assigned to from string literals in very many testcases.  If we change it to `gchar *`, then we get errors all over the place about dropping the `const` qualifier.

Will keep thinking about this...

File: src/common/cockpitwebserver.c [hunk 7]
```diff
@@ -837,16 +821,29 @@ cockpit_web_request_process (CockpitWebRequest *self,
     }
 
   self->original_path = path;
-  self->path = path + self->web_server->url_root->len;
   self->method = method;
   self->headers = headers;
 
+  g_autofree gchar *path_copy = g_strdup (self->original_path + self->web_server->url_root->len);
+  self->path = path_copy;
```
Comment (martinpitt) [inline src/common/cockpitwebserver.c:839]: Note to self: This looks dangerous/wrong at first, but it's just for calling the handler_stream signal handler, and it will be reset to NULL afterwards.

File: src/ws/test-webservice.c [hunk 8]
```diff
@@ -345,6 +356,8 @@ expect_control_message (GBytes *message,
   g_assert_cmpstr (outer_channel, ==, NULL);
   g_free (outer_channel);
 
+  g_printerr ("PAYLOAD %.*s\n", (int)g_bytes_get_size(payload),(char*)g_bytes_get_data(payload, NULL));
```
Comment (martinpitt) [inline src/ws/test-webservice.c:354]: debugging leftover

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! So close, one debug leftover, otherwise LGTM. This trips up the "urlroot with query test" (as you already mentioned on IRC)

**Review by martinpitt — APPROVED** [review]
Danke!
